### PR TITLE
[SPARK-17398][SQL] Fix ClassCastException when querying partitioned JSON table

### PR DESCRIPTION
**Pull request by wypoon**
### What changes were proposed in this pull request?

When querying a partitioned table with format `org.apache.hive.hcatalog.data.JsonSerDe` and more than one task runs in each executor concurrently, the following exception is encountered:

`java.lang.ClassCastException: java.util.ArrayList cannot be cast to org.apache.hive.hcatalog.data.HCatRecord`

The exception occurs in `HadoopTableReader.fillObject`.

`org.apache.hive.hcatalog.data.JsonSerDe#initialize` populates a `cachedObjectInspector` field by calling `HCatRecordObjectInspectorFactory.getHCatRecordObjectInspector`, which is not thread-safe; this `cachedObjectInspector` is returned by `JsonSerDe#getObjectInspector`.

We protect against this Hive bug by synchronizing on an object when we need to call `initialize` on `org.apache.hadoop.hive.serde2.Deserializer` instances (which may be `JsonSerDe` instances). By doing so, the `ObjectInspector` for the `Deserializer` of the partitions of the JSON table and that of the table `SerDe` are the same cached `ObjectInspector` and `HadoopTableReader.fillObject` then works correctly. (If the `ObjectInspector`s are different, then a bug in `HCatRecordObjectInspector` causes an `ArrayList` to be created instead of an `HCatRecord`, resulting in the `ClassCastException` that is seen.)

### Why are the changes needed?

To avoid HIVE-15773 / HIVE-21752.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Tested manually on a cluster with a partitioned JSON table and running a query using more than one core per executor. Before this change, the ClassCastException happens consistently. With this change it does not happen.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 1]
```diff
@@ -252,10 +254,14 @@ class HadoopTableReader(
         partProps.asScala.foreach {
           case (key, value) => props.setProperty(key, value)
         }
-        deserializer.initialize(hconf, props)
+        DeserializerLock.synchronized {
+          deserializer.initialize(hconf, props)
+        }
         // get the table deserializer
         val tableSerDe = localTableDesc.getDeserializerClass.getConstructor().newInstance()
-        tableSerDe.initialize(hconf, localTableDesc.getProperties)
```
Comment (wypoon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: localTableDesc = tableDesc, and tableProperties = tableDesc.getProperties, which was used a little above in this same block, and might as well be used here.

**Comment by dongjoon-hyun** [comment]
Hi, @wypoon . Thank you for making a PR.
Is there a Apache Hive JIRA issue for this?

**Comment by wypoon** [comment]
> Hi, @wypoon . Thank you for making a PR.
> Is there a Apache Hive JIRA issue for this?

HIVE-15773 and HIVE-21752.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 2]
```diff
@@ -132,7 +132,9 @@ class HadoopTableReader(
     val deserializedHadoopRDD = hadoopRDD.mapPartitions { iter =>
       val hconf = broadcastedHadoopConf.value.value
       val deserializer = deserializerClass.getConstructor().newInstance()
-      deserializer.initialize(hconf, localTableDesc.getProperties)
+      DeserializerLock.synchronized {
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Is there any other way to guard against this besides a global lock? that could introduce contention. `deserializer` is a local object here; it accesses some non-thread-safe shared state?

Comment (wypoon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: I don't think there will be a lot of contention. We call `deserializer.initialize` once for each partition.
The race is actually in `HCatRecordObjectInspectorFactory.getHCatRecordObjectInspector` which is called by `JsonSerDe#initialize` (the factory uses a `HashMap` instead of , e.g., a `ConcurrentHashMap` to maintain a cache). We want the same `ObjectInspector` to be returned by the factory cache and set in each `JsonSerDe` instance as its `ObjectInspector`.
Our objective is that there not be more than one task from each executor calling `JsonSerDe#initialize` at the same time.

Comment (srowen) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Fair enough. Is there any way to fix this in the Hadoop implementation? or is it already fixed but we have to deal with it here? 
Can we push the global lock down further rather than making callers know to use it?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 1 above)
Comment (wypoon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Actually, the `tableSerDe` can be instantiated and initialized before the transformation here. Then it will be done just once, instead of repeatedly for each Hive partition.
(The same applies to the `Deserializer` for the table in `makeRDDForTable`.)
I'll make a change.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 2 above)
Comment (wypoon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: HIVE-15773 and HIVE-21752 offer two different fixes for the `HCatRecordObjectInspectorFactory` bug. HIVE-21752 has been merged in Hive 4.0.0, which is not released yet. Even when it is, not many users will be on Hive 4.0.0 for some time, so I think we need to deal with it here.

Comment (wypoon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: On the question of other callers needing to use this global lock:
AFAIK, the only reported problem in Spark with the `HCatRecordObjectInspectorFactory` bug is in reading partitioned `JsonSerDe` tables. In `HadoopTableReader#makeRDDForTable`, we can even do without using the lock; we only really need to use the lock in `HadoopTableReader#makeRDDForPartitionedTable`.
From my search, there are two other places in Spark where we call `Deserializer#initialize`:
1. `HiveTableScanExec`, in initializing a `HadoopTableReader` instance, before `HadoopTableReader#makeRDDForTable` or `HadoopTableReader#makeRDDForPartitionedTable` even get called in `doExec`.
2. `HiveScriptIOSchema`.

I don't think we need to use the lock for 1.
I don't know about 2., but if no problem has been reported due to the race, we can also leave it alone.
In other words, I'm not proposing we guard against the race in the Hive bug everywhere, just in this known case.

**Comment by dongjoon-hyun** [comment]
Thank you, @wypoon . I updated the description of this PR and JIRA links.
> HIVE-15773 and HIVE-21752.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala (hunk 1 above)
Comment (wypoon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: I was mistaken. The SerDe has to be instantiated and initialized within the transformation, as it is not Serializable and cannot be sent in the closure.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala [hunk 3]
```diff
@@ -252,10 +254,14 @@ class HadoopTableReader(
...
         }
-        deserializer.initialize(hconf, props)
+        DeserializerLock.synchronized {
+          deserializer.initialize(hconf, props)
+        }
         // get the table deserializer
         val tableSerDe = localTableDesc.getDeserializerClass.getConstructor().newInstance()
-        tableSerDe.initialize(hconf, localTableDesc.getProperties)
+        DeserializerLock.synchronized {
+          tableSerDe.initialize(hconf, tableProperties)
```
Comment (vanzin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: What you have is ok but you can completely avoid this second serde instance; in the common case both `localDeserializer` and `localTableDesc.getDeserializerClass` will be the same.

Comment (wypoon) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Yes, I did find that to be the case in my repro, that the two were the same class (JsonSerDe). However, the initialize calls on deserializer and on tableSerDe are with potentially different properties (props and tableProperties could differ), so I think I should initialize tableSerDe even if it has the same class as deserializer.

Comment (vanzin) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Ah, ok, missed that whjen reading the code.

**Comment by vanzin** [comment]
Looks good, will merge tomorrow if no comments.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by vanzin** [comment]
Merging to master / 2.4.

**Comment by vanzin** [comment]
(Trivial conflict when merging to 2.4, fixing manually.)

**Comment by vanzin** [comment]
(argh, messed something up when merging to 2.4, will do it manually...)

**Comment by wypoon** [comment]
Thanks @vanzin !

**Comment by vungle-wesleygoi** [comment]
I'm still facing this issue
